src/Workspaces/Remote/ServiceHub/Host/SolutionAssetCache.cs

```diff
@@ -141,39 +154,82 @@ private void ForceGC()
             }
 
             var current = DateTime.UtcNow;
-            using (Logger.LogBlock(FunctionId.AssetStorage_CleanAssets, CancellationToken.None))
+            using (Logger.LogBlock(FunctionId.AssetStorage_CleanAssets, cancellationToken))
             {
-                foreach (var (checksum, entry) in _assets.ToArray())
+                // Ensure that if our remote workspace has a current solution, that we don't purge any items associated
+                // with that solution.
+                PooledHashSet<Checksum>? pinnedChecksums = null;
+                try
+                {
+                    foreach (var (checksum, entry) in _assets)
+                    {
+                        // If not enough time has passed, keep in the cache.
+                        if (current - entry.LastAccessed <= _purgeAfterTimeSpan)
+                            continue;
+
+                        // If this is a checksum we want to pin, do not remove it.
+                        if (pinnedChecksums == null)
+                        {
+                            pinnedChecksums = PooledHashSet<Checksum>.GetInstance();
+                            await AddPinnedChecksumsAsync(pinnedChecksums, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        if (pinnedChecksums.Contains(checksum))
+                            continue;
+
+                        _assets.TryRemove(checksum, out _);
+                    }
+                }
+                finally
                 {
-                    if (current - entry.LastAccessed <= _purgeAfterTimeSpan)
-                        continue;
+                    pinnedChecksums?.Free();
+                }
+            }
+        }
 
-                    // If it fails, we'll just leave it in the asset pool.
-                    _assets.TryRemove(checksum, out var _);
+        private async ValueTask AddPinnedChecksumsAsync(HashSet<Checksum> pinnedChecksums, CancellationToken cancellationToken)
+        {
+            if (_remoteWorkspace is null)
+                return;
+
+            var checksums = await _remoteWorkspace.CurrentSolution.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+
+            Recurse(checksums);
+
+            return;
+
+            void Recurse(ChecksumWithChildren checksums)
+            {
+                pinnedChecksums.Add(checksums.Checksum);
+                foreach (var child in checksums.Children)
+                {
+                    if (child is ChecksumWithChildren childChecksums)
+                        Recurse(childChecksums);
+                    else if (child is Checksum childChecksum)
+                        pinnedChecksums.Add(childChecksum);
                 }
             }
         }
 
         private sealed class Entry
         {
             // mutable field
-            public DateTime LastAccessed;
+            public DateTime LastAccessed = DateTime.UtcNow;
```

Comment:
Seeing this class generally using UTC time is terrifying me if strange things happen when my machine clock resyncs...

Reply:
yeah.  it seems like a great way to have entries that just stick around forever until vs relaunch.  Any tips on something better we can switch to?

Reply:
@CyrusNajmabadi SharedStopwatch? I think memory usage would be the same?

Reply:
could do that.  are those safe for the resync issue?